funsor/adjoint.py

```diff
@@ -2,49 +2,82 @@
 
 
+def _alpha_deconvert(expr):
+    alpha_subs = {name: name.split("__BOUND")[0]
+                  for name in expr.bound if "__BOUND" in name}
+    if not alpha_subs:
+        return tuple(expr._ast_values)
+
+    return alpha_substitute(expr, alpha_subs)
+
+
 class AdjointTape(object):
 
     def __init__(self):
         self.tape = []
+        self._old_interpretation = None
 
     def __call__(self, cls, *args):
-        result = eager(cls, *args)
-        if cls in (Reduce, Contract, Binary, Tensor):
+        result = cls(*args) if self._old_interpretation is None \
+            else self._old_interpretation(cls, *args)
+        if cls in (Reduce, Contract, Binary, Tensor):  # TODO make generic
             self.tape.append((result, cls, args))
         return result
 
+    def __enter__(self):
+        self.tape = []
+        self._old_interpretation = interpreter._INTERPRETATION
+        interpreter.set_interpretation(self)
+        return self
+
+    def __exit__(self, *args):
+        interpreter.set_interpretation(self._old_interpretation)
+        self._old_interpretation = None
+
+    def adjoint(self, red_op, bin_op, root, targets):
+
+        bin_unit = to_funsor(ops.UNITS[bin_op])
+
+        adjoint_values = defaultdict(lambda: bin_unit)
+        multiplicities = defaultdict(lambda: 0)
 
-def adjoint(expr, targets, start=Number(0.)):
+        reached_root = False
+        while self.tape:
+            output, fn, inputs = self.tape.pop()
+            if not reached_root:
+                if output is root:
+                    reached_root = True
+                else:
+                    continue
 
-    adjoint_values = defaultdict(lambda: Number(0.))  # 1 in logspace
-    multiplicities = defaultdict(lambda: 0)
+            # reverse the effects of alpha-renaming
+            with interpretation(lazy):
+                other_subs = {name: name.split("__BOUND")[0] for name in output.inputs if "__BOUND" in name}
+                inputs = _alpha_deconvert(fn(*inputs)(**other_subs))
+                output = type(output)(*_alpha_deconvert(output(**other_subs)))
 
-    tape_recorder = AdjointTape()
-    with interpretation(tape_recorder):
-        adjoint_values[reinterpret(expr)] = start
+            in_adjs = adjoint_ops(fn, red_op, bin_op, adjoint_values[output], *inputs)
+            for v, adjv in in_adjs.items():
+                multiplicities[v] += 1
+                adjoint_values[v] = bin_op(adjoint_values[v], adjv)
 
-    while tape_recorder.tape:
-        output, fn, inputs = tape_recorder.tape.pop()
-        in_adjs = adjoint_ops(fn, adjoint_values[output], output, *inputs)
-        for v, adjv in in_adjs.items():
-            multiplicities[v] += 1
-            adjoint_values[v] = adjoint_values[v] + adjv  # product in logspace
+        target_adjs = {}
+        for v in targets:
+            target_adjs[v] = adjoint_values[v] / multiplicities[v]  # TODO use correct op here with bin_op
```

Comment:
This seems fishy. I'd not expect a `.mean()` but rather a `.sum()` here. Maybe this is caused by the `modulo_total=True` option in Pyro, in which case you could change the test to also compute and divide by the total `Z`.